[explicit-module-boundary-types] Maximum call stack size exceeded error for self-referencing class declaration

<!--
Please don't ignore this template.

If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
Make sure you read through our FAQ before posting.
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md
-->

**Repro**

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.
-->

```YAML
env:
  browser: true
  es2020: true
extends:
  - 'eslint:recommended'
  - 'plugin:@typescript-eslint/recommended'
parser: '@typescript-eslint/parser'
parserOptions:
  ecmaVersion: 11
  sourceType: module
plugins:
  - '@typescript-eslint'
rules: {}
```

```TS
export class A {
  b = A;
}
```

**Expected Result**

eslint does not crash.

**Actual Result**

```
Oops! Something went wrong! :(

ESLint: 7.5.0

RangeError: Maximum call stack size exceeded
Occurred while linting /private/tmp/eslint-repro/index.ts:1
    at Object.get [as AST_NODE_TYPES] (/private/tmp/eslint-repro/node_modules/@typescript-eslint/experimental-utils/dist/index.js:4:67)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:270:43)
    at followReference (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:253:17)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:299:28)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:283:28)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:287:25)
    at followReference (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:253:17)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:299:28)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:283:28)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:287:25)
```

**Additional Info**

<!--
Did eslint throw an exception?

Please run your lint again with the --debug flag, and dump the output below.
i.e. eslint --ext ".ts,.js" src --debug
-->

<details>
  <summary>output with `--debug` option</summary>

```
eslint:cli CLI args: [ './index.ts', '--debug' ] +0ms
  eslint:cli Running on files +3ms
  eslint:config-array-factory Loading JSON config file: /private/tmp/eslint-repro/package.json +0ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/private/tmp/eslint-repro', loose: false } ] +0ms
  eslint:ignore-pattern   processed: { basePath: '/private/tmp/eslint-repro', patterns: [ '/**/node_modules/*' ] } +2ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/private/tmp/eslint-repro', loose: false } ] +0ms
  eslint:ignore-pattern   processed: { basePath: '/private/tmp/eslint-repro', patterns: [ '/**/node_modules/*' ] } +1ms
  eslint:file-enumerator Start to iterate files: [ './index.ts' ] +0ms
  eslint:file-enumerator File: /private/tmp/eslint-repro/index.ts +1ms
  eslint:cascading-config-array-factory Load config files for /private/tmp/eslint-repro. +0ms
  eslint:cascading-config-array-factory No cache found: /private/tmp/eslint-repro. +0ms
  eslint:config-array-factory Loading YAML config file: /private/tmp/eslint-repro/.eslintrc.yml +5ms
  eslint:config-array-factory Config file found: /private/tmp/eslint-repro/.eslintrc.yml +48ms
  eslint:config-array-factory Loading {extends:"eslint:recommended"} relative to /private/tmp/eslint-repro/.eslintrc.yml +0ms
  eslint:config-array-factory Loading JS config file: /private/tmp/eslint-repro/node_modules/eslint/conf/eslint-recommended.js +1ms
  eslint:config-array-factory Loading {extends:"plugin:@typescript-eslint/recommended"} relative to /private/tmp/eslint-repro/.eslintrc.yml +1ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from /private/tmp/eslint-repro/.eslintrc.yml +0ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@3.6.1 (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js) +2ms
  eslint:config-array-factory Plugin /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js loaded in: 407ms +407ms
  eslint:config-array-factory Loading {extends:"./configs/base"} relative to /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js +0ms
  eslint:config-array-factory package.json was not found: Cannot find module './configs/base/package.json'
Require stack:
- /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js +0ms
  eslint:config-array-factory Loaded: ./configs/base (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.js) +0ms
  eslint:config-array-factory Loading JS config file: /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.js +0ms
  eslint:config-array-factory Loading parser "@typescript-eslint/parser" from /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.js +1ms
  eslint:config-array-factory Loaded: @typescript-eslint/parser@3.6.1 (/private/tmp/eslint-repro/node_modules/@typescript-eslint/parser/dist/index.js) +1ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.js +52ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@3.6.1 (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js) +0ms
  eslint:config-array-factory Plugin /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js loaded in: 0ms +0ms
  eslint:config-array-factory Loading {extends:"./configs/eslint-recommended"} relative to /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js +0ms
  eslint:config-array-factory package.json was not found: Cannot find module './configs/eslint-recommended/package.json'
Require stack:
- /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js +0ms
  eslint:config-array-factory Loaded: ./configs/eslint-recommended (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/eslint-recommended.js) +0ms
  eslint:config-array-factory Loading JS config file: /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/eslint-recommended.js +0ms
  eslint:config-array-factory Loading parser "@typescript-eslint/parser" from /private/tmp/eslint-repro/.eslintrc.yml +7ms
  eslint:config-array-factory Loaded: @typescript-eslint/parser@3.6.1 (/private/tmp/eslint-repro/node_modules/@typescript-eslint/parser/dist/index.js) +0ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from /private/tmp/eslint-repro/.eslintrc.yml +0ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@3.6.1 (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js) +0ms
  eslint:config-array-factory Plugin /private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js loaded in: 0ms +0ms
  eslint:cascading-config-array-factory No cache found: /private/tmp. +521ms
  eslint:config-array-factory Config file not found on /private/tmp +0ms
  eslint:cascading-config-array-factory No cache found: /private. +0ms
  eslint:config-array-factory Config file not found on /private +1ms
  eslint:cascading-config-array-factory No cache found: /. +1ms
  eslint:config-array-factory Config file not found on / +0ms
  eslint:rules Loading rule 'constructor-super' (remaining=281) +0ms
  eslint:rules Loading rule 'for-direction' (remaining=280) +2ms
  eslint:rules Loading rule 'getter-return' (remaining=279) +1ms
  eslint:rules Loading rule 'no-async-promise-executor' (remaining=278) +2ms
  eslint:rules Loading rule 'no-case-declarations' (remaining=277) +1ms
  eslint:rules Loading rule 'no-class-assign' (remaining=276) +1ms
  eslint:rules Loading rule 'no-compare-neg-zero' (remaining=275) +1ms
  eslint:rules Loading rule 'no-cond-assign' (remaining=274) +1ms
  eslint:rules Loading rule 'no-const-assign' (remaining=273) +2ms
  eslint:rules Loading rule 'no-constant-condition' (remaining=272) +1ms
  eslint:rules Loading rule 'no-control-regex' (remaining=271) +1ms
  eslint:rules Loading rule 'no-debugger' (remaining=270) +1ms
  eslint:rules Loading rule 'no-delete-var' (remaining=269) +0ms
  eslint:rules Loading rule 'no-dupe-args' (remaining=268) +1ms
  eslint:rules Loading rule 'no-dupe-class-members' (remaining=267) +1ms
  eslint:rules Loading rule 'no-dupe-else-if' (remaining=266) +0ms
  eslint:rules Loading rule 'no-dupe-keys' (remaining=265) +1ms
  eslint:rules Loading rule 'no-duplicate-case' (remaining=264) +0ms
  eslint:rules Loading rule 'no-empty' (remaining=263) +1ms
  eslint:rules Loading rule 'no-empty-character-class' (remaining=262) +2ms
  eslint:rules Loading rule 'no-empty-pattern' (remaining=261) +0ms
  eslint:rules Loading rule 'no-ex-assign' (remaining=260) +1ms
  eslint:rules Loading rule 'no-extra-boolean-cast' (remaining=259) +1ms
  eslint:rules Loading rule 'no-extra-semi' (remaining=258) +1ms
  eslint:rules Loading rule 'no-fallthrough' (remaining=257) +1ms
  eslint:rules Loading rule 'no-func-assign' (remaining=256) +1ms
  eslint:rules Loading rule 'no-global-assign' (remaining=255) +1ms
  eslint:rules Loading rule 'no-import-assign' (remaining=254) +2ms
  eslint:rules Loading rule 'no-inner-declarations' (remaining=253) +1ms
  eslint:rules Loading rule 'no-invalid-regexp' (remaining=252) +1ms
  eslint:rules Loading rule 'no-irregular-whitespace' (remaining=251) +1ms
  eslint:rules Loading rule 'no-misleading-character-class' (remaining=250) +2ms
  eslint:rules Loading rule 'no-mixed-spaces-and-tabs' (remaining=249) +4ms
  eslint:rules Loading rule 'no-new-symbol' (remaining=248) +1ms
  eslint:rules Loading rule 'no-obj-calls' (remaining=247) +1ms
  eslint:rules Loading rule 'no-octal' (remaining=246) +1ms
  eslint:rules Loading rule 'no-prototype-builtins' (remaining=245) +1ms
  eslint:rules Loading rule 'no-redeclare' (remaining=244) +1ms
  eslint:rules Loading rule 'no-regex-spaces' (remaining=243) +1ms
  eslint:rules Loading rule 'no-self-assign' (remaining=242) +1ms
  eslint:rules Loading rule 'no-setter-return' (remaining=241) +2ms
  eslint:rules Loading rule 'no-shadow-restricted-names' (remaining=240) +1ms
  eslint:rules Loading rule 'no-sparse-arrays' (remaining=239) +1ms
  eslint:rules Loading rule 'no-this-before-super' (remaining=238) +0ms
  eslint:rules Loading rule 'no-undef' (remaining=237) +1ms
  eslint:rules Loading rule 'no-unexpected-multiline' (remaining=236) +1ms
  eslint:rules Loading rule 'no-unreachable' (remaining=235) +1ms
  eslint:rules Loading rule 'no-unsafe-finally' (remaining=234) +2ms
  eslint:rules Loading rule 'no-unsafe-negation' (remaining=233) +1ms
  eslint:rules Loading rule 'no-unused-labels' (remaining=232) +1ms
  eslint:rules Loading rule 'no-unused-vars' (remaining=231) +1ms
  eslint:rules Loading rule 'no-useless-catch' (remaining=230) +1ms
  eslint:rules Loading rule 'no-useless-escape' (remaining=229) +1ms
  eslint:rules Loading rule 'no-with' (remaining=228) +1ms
  eslint:rules Loading rule 'require-yield' (remaining=227) +1ms
  eslint:rules Loading rule 'use-isnan' (remaining=226) +1ms
  eslint:rules Loading rule 'valid-typeof' (remaining=225) +1ms
  eslint:rules Loading rule 'no-var' (remaining=224) +2ms
  eslint:rules Loading rule 'prefer-const' (remaining=223) +1ms
  eslint:rules Loading rule 'prefer-rest-params' (remaining=222) +2ms
  eslint:rules Loading rule 'prefer-spread' (remaining=221) +1ms
  eslint:rules Loading rule 'no-array-constructor' (remaining=220) +5ms
  eslint:rules Loading rule 'no-empty-function' (remaining=219) +1ms
  eslint:cascading-config-array-factory Configuration was determined: ConfigArray(7) [ { type: 'config', name: 'DefaultIgnorePattern', filePath: '', criteria: null, env: undefined, globals: undefined, ignorePattern: IgnorePattern { patterns: [Array], basePath: '/private/tmp/eslint-repro', loose: false }, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '.eslintrc.yml » eslint:recommended', filePath: '/private/tmp/eslint-repro/node_modules/eslint/conf/eslint-recommended.js', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'constructor-super': 'error', 'for-direction': 'error', 'getter-return': 'error', 'no-async-promise-executor': 'error', 'no-case-declarations': 'error', 'no-class-assign': 'error', 'no-compare-neg-zero': 'error', 'no-cond-assign': 'error', 'no-const-assign': 'error', 'no-constant-condition': 'error', 'no-control-regex': 'error', 'no-debugger': 'error', 'no-delete-var': 'error', 'no-dupe-args': 'error', 'no-dupe-class-members': 'error', 'no-dupe-else-if': 'error', 'no-dupe-keys': 'error', 'no-duplicate-case': 'error', 'no-empty': 'error', 'no-empty-character-class': 'error', 'no-empty-pattern': 'error', 'no-ex-assign': 'error', 'no-extra-boolean-cast': 'error', 'no-extra-semi': 'error', 'no-fallthrough': 'error', 'no-func-assign': 'error', 'no-global-assign': 'error', 'no-import-assign': 'error', 'no-inner-declarations': 'error', 'no-invalid-regexp': 'error', 'no-irregular-whitespace': 'error', 'no-misleading-character-class': 'error', 'no-mixed-spaces-and-tabs': 'error', 'no-new-symbol': 'error', 'no-obj-calls': 'error', 'no-octal': 'error', 'no-prototype-builtins': 'error', 'no-redeclare': 'error', 'no-regex-spaces': 'error', 'no-self-assign': 'error', 'no-setter-return': 'error', 'no-shadow-restricted-names': 'error', 'no-sparse-arrays': 'error', 'no-this-before-super': 'error', 'no-undef': 'error', 'no-unexpected-multiline': 'error', 'no-unreachable': 'error', 'no-unsafe-finally': 'error', 'no-unsafe-negation': 'error', 'no-unused-labels': 'error', 'no-unused-vars': 'error', 'no-useless-catch': 'error', 'no-useless-escape': 'error', 'no-with': 'error', 'require-yield': 'error', 'use-isnan': 'error', 'valid-typeof': 'error' }, settings: undefined }, { type: 'config', name: '.eslintrc.yml » plugin:@typescript-eslint/recommended » ./configs/base', filePath: '/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.js', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: { error: null, filePath: '/private/tmp/eslint-repro/node_modules/@typescript-eslint/parser/dist/index.js', id: '@typescript-eslint/parser', importerName: '.eslintrc.yml » plugin:@typescript-eslint/recommended » ./configs/base', importerPath: '/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/base.js' }, parserOptions: { sourceType: 'module' }, plugins: { '@typescript-eslint': [Object] }, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '.eslintrc.yml » plugin:@typescript-eslint/recommended » ./configs/eslint-recommended', filePath: '/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/eslint-recommended.js', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '.eslintrc.yml » plugin:@typescript-eslint/recommended » ./configs/eslint-recommended#overrides[0]', filePath: '/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/configs/eslint-recommended.js', criteria: { includes: [Array], excludes: null, basePath: '/private/tmp/eslint-repro' }, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'constructor-super': 'off', 'getter-return': 'off', 'no-const-assign': 'off', 'no-dupe-args': 'off', 'no-dupe-class-members': 'off', 'no-dupe-keys': 'off', 'no-func-assign': 'off', 'no-import-assign': 'off', 'no-new-symbol': 'off', 'no-obj-calls': 'off', 'no-redeclare': 'off', 'no-setter-return': 'off', 'no-this-before-super': 'off', 'no-undef': 'off', 'no-unreachable': 'off', 'no-unsafe-negation': 'off', 'no-var': 'error', 'prefer-const': 'error', 'prefer-rest-params': 'error', 'prefer-spread': 'error', 'valid-typeof': 'off' }, settings: undefined }, { type: 'config', name: '.eslintrc.yml » plugin:@typescript-eslint/recommended', filePath: '/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { '@typescript-eslint/adjacent-overload-signatures': 'error', '@typescript-eslint/ban-ts-comment': 'error', '@typescript-eslint/ban-types': 'error', '@typescript-eslint/explicit-module-boundary-types': 'warn', 'no-array-constructor': 'off', '@typescript-eslint/no-array-constructor': 'error', 'no-empty-function': 'off', '@typescript-eslint/no-empty-function': 'error', '@typescript-eslint/no-empty-interface': 'error', '@typescript-eslint/no-explicit-any': 'warn', '@typescript-eslint/no-extra-non-null-assertion': 'error', 'no-extra-semi': 'off', '@typescript-eslint/no-extra-semi': 'error', '@typescript-eslint/no-inferrable-types': 'error', '@typescript-eslint/no-misused-new': 'error', '@typescript-eslint/no-namespace': 'error', '@typescript-eslint/no-non-null-asserted-optional-chain': 'error', '@typescript-eslint/no-non-null-assertion': 'warn', '@typescript-eslint/no-this-alias': 'error', 'no-unused-vars': 'off', '@typescript-eslint/no-unused-vars': 'warn', '@typescript-eslint/no-var-requires': 'error', '@typescript-eslint/prefer-as-const': 'error', '@typescript-eslint/prefer-namespace-keyword': 'error', '@typescript-eslint/triple-slash-reference': 'error' }, settings: undefined }, { type: 'config', name: '.eslintrc.yml', filePath: '/private/tmp/eslint-repro/.eslintrc.yml', criteria: null, env: { browser: true, es2020: true }, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: { error: null, filePath: '/private/tmp/eslint-repro/node_modules/@typescript-eslint/parser/dist/index.js', id: '@typescript-eslint/parser', importerName: '.eslintrc.yml', importerPath: '/private/tmp/eslint-repro/.eslintrc.yml' }, parserOptions: { ecmaVersion: 11, sourceType: 'module' }, plugins: { '@typescript-eslint': [Object] }, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: {}, settings: undefined } ] on /private/tmp/eslint-repro +79ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/private/tmp/eslint-repro', loose: false } ] +620ms
  eslint:ignore-pattern   processed: { basePath: '/private/tmp/eslint-repro', patterns: [ '/**/node_modules/*' ] } +1ms
  eslint:ignore-pattern Check {
  filePath: '/private/tmp/eslint-repro/index.ts',
  dot: false,
  relativePath: 'index.ts',
  result: false
} +0ms
  eslint:cli-engine Lint /private/tmp/eslint-repro/index.ts +0ms
  eslint:linter Linting code for /private/tmp/eslint-repro/index.ts (pass 1) +0ms
  eslint:linter Verify +1ms
  eslint:linter With ConfigArray: /private/tmp/eslint-repro/index.ts +0ms
  eslint:linter An error occurred while traversing +42ms
  eslint:linter Filename: /private/tmp/eslint-repro/index.ts +1ms
  eslint:linter Line: 1 +0ms
  eslint:linter Parser Options: {
  ecmaVersion: 11,
  sourceType: 'module',
  ecmaFeatures: { globalReturn: false }
} +0ms
  eslint:linter Parser Path: /private/tmp/eslint-repro/node_modules/@typescript-eslint/parser/dist/index.js +0ms
  eslint:linter Settings: {} +0ms

Oops! Something went wrong! :(

ESLint: 7.5.0

RangeError: Maximum call stack size exceeded
Occurred while linting /private/tmp/eslint-repro/index.ts:1
    at Object.get [as AST_NODE_TYPES] (/private/tmp/eslint-repro/node_modules/@typescript-eslint/experimental-utils/dist/index.js:4:67)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:270:43)
    at followReference (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:253:17)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:299:28)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:283:28)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:287:25)
    at followReference (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:253:17)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:299:28)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:283:28)
    at checkNode (/private/tmp/eslint-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/explicit-module-boundary-types.js:287:25)
```
</details>

It seems that we can avoid the error by disabling the `explicit-module-boundary-types` rule.

**Versions**

| package                                 | version |
| --------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.6.1` |
| `@typescript-eslint/experimental-utils` | `3.6.1` |
| `eslint` | `7.5.0` |
| `TypeScript`                            | `3.9.7` |
| `node`                                  | `13.8.0` |
| `npm`                                   | `6.14.6` |

--- a/packages/eslint-plugin/src/rules/explicit-module-boundary-types.ts
+++ b/packages/eslint-plugin/src/rules/explicit-module-boundary-types.ts
@@ -96,6 +96,10 @@ export default util.createRule<Options, MessageIds>({
     // tracks functions that were found whilst traversing
     const foundFunctions: FunctionNode[] = [];
 
+    // all nodes visited, avoids infinite recursion for cyclic references
+    // (such as class member referring to itself)
+    const alreadyVisited = new Set<TSESTree.Node>();
+
     /*
     # How the rule works:
 
@@ -106,7 +110,6 @@ export default util.createRule<Options, MessageIds>({
     After it's finished traversing the AST, it then iterates through the list of found functions, and checks to see if
     any of them are part of a higher-order function
     */
-    const alreadyVisited = new Set<TSESTree.Node>();
 
     return {
       ExportDefaultDeclaration(node): void {
@@ -310,6 +313,7 @@ export default util.createRule<Options, MessageIds>({
         }
       }
     }
+
     function checkNode(node: TSESTree.Node | null): void {
       if (node == null || alreadyVisited.has(node)) {
         return;

--- a/packages/eslint-plugin/src/rules/explicit-module-boundary-types.ts
+++ b/packages/eslint-plugin/src/rules/explicit-module-boundary-types.ts
@@ -106,6 +106,7 @@ export default util.createRule<Options, MessageIds>({
     After it's finished traversing the AST, it then iterates through the list of found functions, and checks to see if
     any of them are part of a higher-order function
     */
+    const alreadyVisited = new Set<TSESTree.Node>();
 
     return {
       ExportDefaultDeclaration(node): void {
@@ -309,11 +310,11 @@ export default util.createRule<Options, MessageIds>({
         }
       }
     }
-
     function checkNode(node: TSESTree.Node | null): void {
-      if (node == null) {
+      if (node == null || alreadyVisited.has(node)) {
         return;
       }
+      alreadyVisited.add(node);
 
       switch (node.type) {
         case AST_NODE_TYPES.ArrowFunctionExpression:

--- a/packages/eslint-plugin/tests/rules/explicit-module-boundary-types.test.ts
+++ b/packages/eslint-plugin/tests/rules/explicit-module-boundary-types.test.ts
@@ -656,6 +656,11 @@ export declare class Foo {
   set time(seconds: number);
 }
     `,
+    `
+class A {
+  b = A;
+}
+    `,
   ],
   invalid: [
     {

--- a/packages/eslint-plugin/tests/rules/explicit-module-boundary-types.test.ts
+++ b/packages/eslint-plugin/tests/rules/explicit-module-boundary-types.test.ts
@@ -657,7 +657,7 @@ export declare class Foo {
 }
     `,
     `
-class A {
+export class A {
   b = A;
 }
     `,

